source/common/http/http1/codec_impl.cc

```diff
@@ -376,16 +377,80 @@ void ServerConnectionImpl::onEncodeComplete() {
   }
 }
 
+
+  // handlePath computes the correct :authority and :path headers based on 7230#5.7 and 7230#6
+  //
+  bool
+  ServerConnectionImpl::handlePath(HeaderMapImplPtr& headers, bool is_connect) {
+  // TODO(mattwoodyard) - check configuration option if forward proxy is enabled
+  // In practice there will only ever be one scheme delivered for forward proxy
+  // support
+    HeaderString path(Headers::get().Path);
+
+    // The url is relative nothing to do here
+    if (active_request_->request_url_.c_str()[0] == '/' || active_request_->request_url_.c_str()[0] == '*') {
+      headers->addViaMove(std::move(path), std::move(active_request_->request_url_));
+      return true;
+    }
+
+    if (!codec_settings_.allow_connect_ && !codec_settings_.allow_absolute_url_) {
+      return true;
+    }
+
+    struct http_parser_url u;
+    http_parser_url_init(&u);
+    int result = http_parser_parse_url(active_request_->request_url_.buffer(), active_request_->request_url_.size(), is_connect, &u);
+    if (result != HPE_OK) {
+      ENVOY_CONN_LOG(trace, "invalid uri in request", connection_);
+      return false;
+    } else {
+      if ((u.field_set & UF_HOST) == UF_HOST && (u.field_set & UF_SCHEMA) == UF_SCHEMA) {
+        // RFC7230#5.7
+        // When a proxy receives a request with an absolute-form of
+        // request-target, the proxy MUST ignore the received Host header field
+        // (if any) and instead replace it with the host information of the
+        // request-target.  A proxy that forwards such a request MUST generate a
+        // new Host field-value based on the received request-target rather than
+        // forward the received Host field-value.
+
+        // Insert the host header, this will later be converted to :authority
+        std::string new_host(active_request_->request_url_.c_str() + u.field_data[UF_HOST].off, u.field_data[UF_HOST].len);
+        headers->insertHost().value(new_host);
+
+        // RFC allows the absolute-uri to not end in /, but the absolute path form
+        // must start with /
+        if ((u.field_set & UF_HOST) == UF_PATH) {
+          HeaderString new_path;
+          new_path.setCopy(active_request_->request_url_.c_str() + u.field_data[UF_PATH].off,
+                           active_request_->request_url_.size() - u.field_data[UF_PATH].off);
+          headers->addViaMove(std::move(path), std::move(new_path));
+        } else {
+          HeaderString new_path;
+          new_path.setCopy("/", 1);
+          headers->addViaMove(std::move(path), std::move(new_path));
+        }
+
+        // The existing code clears the request_url_ via move
+        // TODO(mattwoodyard) Is this neccessary, just cargo culted
```

Comment:
If nothing else it's nice to clean up memory as you're done with it :-)
You could even leave the original ASSERT in the code base to verify it's done on all paths.